packages/react-router/src/useBlocker.tsx

```diff
@@ -3,25 +3,62 @@ import { useRouter } from './useRouter'
 import type { BlockerFn } from '@tanstack/history'
 import type { ReactNode } from './route'
 
+type Resolver = {
+  allow: () => void
+  deny: () => void
+}
+
```

Comment:
naming: should we maybe consider prior art such as https://reactrouter.com/en/main/hooks/use-blocker ?

Reply:
react-router also has a `state` param, do we need it?

Reply:
I also kinda like `currentLocation` and `nextLocation`

Reply:
I like the `.proceed()` and `.reset()` methods, and especially the `state` value. Though, I'd rename `state` to `status` that'd be more inline with Query.

Reply:
Including `currentLocation` and `nextLocation` also sounds like a good idea!

It'll require the `useBlocker` hook to be aware of the `nextLocation`, which may require the router to keep track of the blockers (which it doesn't do now) since the next location is most easily available in `router.buildAndCommitLocation`.

I'm not opposed to going forward with this PR without `currentLocation` and `nextLocation`, and then adding it in with some strong tests in place.

Reply:
`state` would be helpful: in the example case, it'd avoid the need to manually set the state (and also renders the blockerFn property unnecessary (even though it's still mandatory to provide, but maybe that can change?)

+1 on the currentLocation/nextLocation, but that probably requires changes in the history package ([here](https://github.com/TanStack/router/blob/main/packages/history/src/index.ts#L78)).

Also, considering the similarity of other hooks between the two libraries, it might be a good idea to end up with something that has a similar usage, so `.proceed()` and .`reset()` could be better suited.

Reply:
Speaking of the state/status, do you manually want to handle it internally on useBlocker or should be something at a router level?

Reply:
Currently, since blockers have no effect on the actual Router, and directly interface with the history package, I'd keep this state local to the `useBlocker` hook and component.

Reply:
With status exposed seems like `blockerFn` might become optional, I just pushed a proposal. What do you think?

Reply:
> With status exposed seems like `blockerFn` might become optional, I just pushed a proposal and updated the example. What do you think?

🚨 The object syntax would make this a breaking change.